FINERACT-1391:AccountNumberPreferencesPrefixEight

rrpawar96 opened the pull request:
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).


## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

bharathcgowda commented:
@rrpawar96  thank you for sending this fix, could you please check and fix the travis issue? 

@ptuomola  @vidakovic @thesmallstar  It would be really helpful if any one of you do the code review.

vidakovic commented:
btw: I have a fix for the Travis issue... will submit later

rrpawar96 commented:
Yes, it seems the Travis build is been fixed by this PR #1981, let me also try again by rebasing.

ptuomola reviewed (changes requested):
@rrpawar96 this looks like a good feature - but why are we hardcoding the prefix to be 8's only? Would this not be much more flexible if the user could specify any string (numeric?) to be used as prefix?

rrpawar96 committed:
FINERACT-1391:AccountNumberPrefixPreference

rrpawar96 commented:
Hi @ptuomola, The comments has been addressed, please re-review the PR.
Two param digits and prefixCharacter, 'prefixCharacter' will be taken to assign the prefix which will be respective length which will stored in the 'digits' param

francisguchie commented:
@ptuomola @rrpawar96  & @vidakovic 

I would like to do a functional test on this please advise

ikimbra commented:
@ptuomola I see the issue has now been fixed.

francisguchie commented:
@rrpawar96  @bharathcgowda  @ptuomola 
I pulled this your branch and did a functional test - It works well for me 

> @ptuomola I see the issue has now been fixed.

@ikimbra  did you do a test please advise

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/data/AccountNumberFormatDataValidator.java:
```diff
@@ -84,6 +86,19 @@ public void validateForCreate(final String json) {
             DataValidatorBuilder dataValidatorForValidatingPrefixType = baseDataValidator.reset()
                     .parameter(AccountNumberFormatConstants.prefixTypeParamName).value(prefixType).notNull().integerGreaterThanZero();
 
+            if (prefixType == 401) {
```
What's 401? Instead of using a "magic number" can we use the correct constant for this?

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormat.java:
```diff
@@ -37,21 +37,40 @@
         if (prefixType != null) {
             this.prefixEnum = prefixType.getValue();
         }
+        if (prefixType.getValue() == 401) {
```
Same as above

ptuomola replied:
Why would we only set these variables for a specific type? What people expect is that the constructor will set the parameters that are passed in... what's the harm of just setting these every time?

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/service/AccountNumberFormatWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -68,7 +68,15 @@ public CommandProcessingResult createAccountNumberFormat(JsonCommand command) {
-            AccountNumberFormat accountNumberFormat = new AccountNumberFormat(entityAccountType, accountNumberPrefixType);
+            Integer digits = 0;
+            String prefixCharacter = null;
+            if (prefixTypeId == 401) {
```
As above

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -134,11 +147,50 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
                 prefix = prefix.substring(0, Math.min(prefix.length(), 10));
             }
 
+            if (accountNumberPrefixType.getValue() == 401) {
```
Why do we need this block? We've already determined the prefix above....  can't we just use the length of the prefix to determine how much we need to pad the account number? Why do we need to retrieve the length again?

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -134,11 +147,50 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
+    private Map<String, String> prefixEight(Map<String, String> propertyMap, String accountNumber, Integer accountMaxLength,
+            AccountNumberFormat accountNumberFormat) {
+        AccountNumberFormat accountNumberByPrefix = this.accountNumberFormatRepository
+                .findOneByPrefixEnumAccountTypeEnum(accountNumberFormat.getAccountTypeEnum(), 401);
```
Why do we need to look up again the AccountNumberFormat? Should we not use the one that we've been passed in as an argument, rather than looking up a new one?

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -134,11 +147,50 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
+            AccountNumberFormat accountNumberFormat) {
+        AccountNumberFormat accountNumberByPrefix = this.accountNumberFormatRepository
+                .findOneByPrefixEnumAccountTypeEnum(accountNumberFormat.getAccountTypeEnum(), 401);
+        Integer digits = accountNumberByPrefix.getDigits();
```
Do we need to pass in digits as a separate argument and store it separately in the database? Could we not infer the length of the prefix from the length of the prefix string? Ie if we want the prefix to be 3 characters of "123" we just store "123" as the prefix string.

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -134,11 +147,50 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
+        Integer numberLength = accountMaxLength - digits;
+
+        if (Integer.valueOf(propertyMap.get(ID).length()) == numberLength + 1) { // 888-999999
+            prefix = prefix.substring(0, prefix.length() - 1);
```
This logic seems to be a) hardcoded to expect a specific length for prefix and digits b) only handle specific cases of length/prefix length. Can this not be made generic so that we automatically work out the length of the number / prefix based on the lengths of the inputs?

rrpawar96 commented:
Thanks, @ptuomola for reviewing, I have addressed all comments, Requesting for re-review.
Removed the use of Digits and integer, used the length of Prefix and constant name instead, created new generic logic for creating Prefix and tested.

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormatRepository.java:
```diff
@@ -28,6 +28,12 @@
     String FIND_ACCOUNT_NUMBER_FORMAT_FOR_ENTITY = "select anf from  AccountNumberFormat anf where anf.accountTypeEnum = :accountTypeEnum";
 
+    String FIND_ACCOUNT_NUMBER_FORMAT_FOR_ENTITY_BY_PREFIXENUM_ACCOUNTTYPEENUM = "select anf from  AccountNumberFormat anf where anf.accountTypeEnum = :accountTypeEnum and anf.prefixEnum = :prefixEnum";
+
```
Is this query still needed? I don't see it being called anywhere

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormatRepository.java:
```diff
@@ -28,6 +28,12 @@
     AccountNumberFormat findOneByAccountTypeEnum(@Param("accountTypeEnum") Integer accountTypeEnum);
+
+    @Query(FIND_ACCOUNT_NUMBER_FORMAT_FOR_ENTITY_BY_PREFIXENUM_ACCOUNTTYPEENUM)
+    AccountNumberFormat findOneByPrefixEnumAccountTypeEnum(@Param("accountTypeEnum") Integer accountTypeEnum,
```
Same as above

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -70,6 +77,7 @@ public String generate(Loan loan, AccountNumberFormat accountNumberFormat) {
         Map<String, String> propertyMap = new HashMap<>();
         propertyMap.put(ID, loan.getId().toString());
         propertyMap.put(OFFICE_NAME, loan.getOffice().getName());
+        propertyMap.put(ENTITY_TYPE, "loan");
```
Why is this change needed?

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -78,6 +86,7 @@ public String generate(SavingsAccount savingsAccount, AccountNumberFormat accoun
         Map<String, String> propertyMap = new HashMap<>();
         propertyMap.put(ID, savingsAccount.getId().toString());
         propertyMap.put(OFFICE_NAME, savingsAccount.office().getName());
+        propertyMap.put(ENTITY_TYPE, "savingsAccount");
```
Same as above

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -101,10 +111,9 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
         }
 
-        String accountNumber = StringUtils.leftPad(propertyMap.get(ID), accountMaxLength, '0');
+        accountNumber = StringUtils.leftPad(accountNumber, accountMaxLength, '0');
```
Why do we need to initialise this again? It was initialise already in line 103...

ptuomola commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java:
```diff
@@ -133,12 +146,38 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
         return accountNumber;
     }
 
+    private Map<String, String> prefixEight(Map<String, String> propertyMap, String accountNumber, Integer accountMaxLength,
```
Should we call this function "generatePrefix" rather than "prefixEight"?

ptuomola reviewed (changes requested):
Looks great @rrpawar96 - thanks for making the changes! A few more comments - hope you can address these, and then I think it should be ready for merging...

rrpawar96 commented:
@ptuomola Comment has been addressed, Removed unused properties.

ptuomola reviewed (approved):
LGTM... will wait until the build passes and then merge this
